content-testing/schemas.js

```diff
@@ -0,0 +1,343 @@
+  timestampsArrayValidator,
+  urlOrRelativeLinkValidator
+} = require('./validators');
+
+const baseVideosSchema = strictObject({
+  title: string().required(),
+  description: string().required(),
+
+  videoId: youtubeIdValidator.when(['parts'], (parts, schema) => {
+    // videoId required if the `parts` array is missing or empty
+    return parts && parts.length > 0 ? schema : schema.required();
+  }),
+
+  nebulaSlug: nonUrlStringValidator,
+  date: dateRangeValidator.required(),
+
+  languages: array(string().required()).required(), // the git videos don't have languages so we can't enforce .min(1)
+  topics: array(string().required()).min(1).required(),
+
+  canContribute: boolean().required(),
+
+  parts: array(
+    strictObject({
+      title: string().required(),
+      videoId: youtubeIdValidator.required(),
+      nebulaSlug: nonUrlStringValidator,
+      timestamps: timestampsArrayValidator.min(1).required()
+    }).required()
+  ).min(2), // if we need parts, should have at least 2
+
+  timestamps: timestampsArrayValidator.required(),
+
+  codeExamples: array(
+    strictObject({
+      title: string().required(),
+      description: string().required(),
+
+      image: string()
+        .test(
+          'relative image',
+          'Should be a relative image filename with `.png` or `.jpg` extension',
+          (value) => {
+            if (!value) return true;
+
+            const properFormat = /^\S+\.(?:png|jpg)$/.test(value);
+            if (!properFormat) return false;
+
+            return true;
+          }
+        )
+        .test(
+          'image file exists',
+          'Image file should exist on disk',
+          (value, context) => {
+            if (!value) return true;
+
+            // extract the JSON filepath passed through the context
+            const filepath = context.options.context.filepath;
+
+            // these image files live in {video-dir}/images/{image}
+            const imagepath = path.join(
+              path.dirname(filepath),
+              'images',
+              value
+            );
+            if (!existsSync(imagepath)) return false;
+
+            return true;
+          }
+        ), // not required, falls back to {video-dir}/index.jpg
+
+      urls: strictObject({
+        p5: string().url(),
+        processing: string().url(),
+        node: string().url(),
+        other: string().url()
+      })
+        // all keys are optional but we require at least one to be set
+        .test(
+          'one object key required',
+          'At least one key is required for this object',
+          (value) => Object.keys(value).length > 0
+        )
+        .required()
+    })
+  ),
+
+  groupLinks: array(
+    strictObject({
+      title: string().required(),
+      links: array(
+        strictObject({
+          title: string().required(),
+          url: urlOrRelativeLinkValidator.required(), // TODO ref check instead of regex pattern matching for internal links?
```

Comment:
This might be a good idea, the yt-desc script kind of checks for this, as it tries to convert all relative links to youtube links. Though here we'd need to check for other kinds of links rather than just video/track pages.

https://github.com/CodingTrain/thecodingtrain.com/blob/00545fd2621645b508780c5d14a7f8df5ac3bc5a/node-scripts/yt-description.mjs#L186-L228

Here's a bad url I just found:
https://github.com/CodingTrain/thecodingtrain.com/blob/00545fd2621645b508780c5d14a7f8df5ac3bc5a/content/videos/webgl/6-createGraphics/index.json#L63-L64

Reply:
That was aspirational when I initially wrote this comment, but yeah! We already have all the site slugs in sets, it's probably pretty easy at this point to make a new validator that would handle this. I'll take a stab at it later today, thanks for reminding me.

Reply:
Done! Caught a second bad link in there.